pkg/secrets/check_rights_windows.go

```diff
@@ -117,3 +115,150 @@ func checkRights(filename string, allowGroupExec bool) error {
+		2, // administrators group has 2 valid subauths
+		windows.SECURITY_BUILTIN_DOMAIN_RID,
+		windows.DOMAIN_ALIAS_RID_ADMINS,
+		0, 0, 0, 0, 0, 0,
+		&administrators)
+	return administrators, err
+}
+
+func getDDAgentUserSID() (*windows.SID, error) {
+	k, err := registry.OpenKey(registry.LOCAL_MACHINE, `SOFTWARE\Datadog\Datadog Agent`, registry.QUERY_VALUE)
+	if err != nil {
+		return nil, fmt.Errorf("could not open installer registry key: %s", err)
+	}
+	defer k.Close()
+
+	user, _, err := k.GetStringValue("installedUser")
+	if err != nil {
+		return nil, fmt.Errorf("could not read installedUser in registry: %s", err)
+	}
+
+	domain, _, err := k.GetStringValue("installedDomain")
+	if err != nil {
+		return nil, fmt.Errorf("could not read installedDomain in registry: %s", err)
+	}
+
+	sid, _, _, err := windows.LookupSID(domain, user)
+	return sid, err
+}
+
+var checkConfigRights = func(filename string) error {
+	if _, err := os.Stat(filename); err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("config file %s does not exist", filename)
+		}
+	}
+
+	fileDacl, err := getACL(filename)
+	if err != nil {
+		return fmt.Errorf("could not query ACLs for %s: %s", filename, err)
+	}
+
+	var aclSizeInfo winutil.AclSizeInformation
+	err = winutil.GetAclInformation(fileDacl, &aclSizeInfo, winutil.AclSizeInformationEnum)
+	if err != nil {
+		return fmt.Errorf("could not query ACLs for %s: %s", filename, err)
+	}
+
+	// create the sids that are acceptable to us (local system account and
+	// administrators group)
+	localSystem, err := getLocalSystemSID()
+	if err != nil {
+		return fmt.Errorf("could not query Local System SID: %s", err)
+	}
+	defer windows.FreeSid(localSystem)
+
+	administrators, err := getAdministratorsSID()
+	if err != nil {
+		return fmt.Errorf("could not query Administrator SID: %s", err)
+	}
+	defer windows.FreeSid(administrators)
+
+	for i := uint32(0); i < aclSizeInfo.AceCount; i++ {
+		var pAce *winutil.AccessAllowedAce
+		if err := winutil.GetAce(fileDacl, i, &pAce); err != nil {
+			return fmt.Errorf("Could not query a ACE on %s: %s", filename, err)
+		}
+
+		compareSid := (*windows.SID)(unsafe.Pointer(&pAce.SidStart))
+		compareIsLocalSystem := windows.EqualSid(compareSid, localSystem)
+		compareIsAdministrators := windows.EqualSid(compareSid, administrators)
+
+		if pAce.AceType == winutil.ACCESS_DENIED_ACE_TYPE {
+			return fmt.Errorf("invalid permissions for config file '%s': explicit DENY not supported", filename)
+		}
+
+		if pAce.AceType == winutil.ACCESS_ALLOWED_ACE_TYPE {
+			allowedAccountForWrite := compareIsLocalSystem || compareIsAdministrators
+			hasOnlyAllowForRead := pAce.AccessMask&^(windows.FILE_GENERIC_READ) == 0
+			if !allowedAccountForWrite && !hasOnlyAllowForRead {
+				return fmt.Errorf("invalid permissions for config file '%s': users/groups other than LOCAL_SYSTEM and Administrators have rights on it", filename)
+			}
+		}
+	}
```

Comment:
This block looks very similar to L85+ lines. Can we extract commonality to a separate func?

Reply:
I thought about it - my concern is that it would obscure reviewing this code. I tried to extract other shared bits like retrieval of SIDs.